File: 周赛/leetcode每日一题/_03_24/661. 图片平滑器.go
package _03_24

func imageSmoother(img [][]int) [][]int {
	m, n := len(img), len(img[0])
	presum, ans := make([][]int, m+1), make([][]int, m)
	presum[0] = make([]int, n+1)
	for i := 0; i < m; i++ {
		presum[i+1] = make([]int, n+1)
		ans[i] = make([]int, n)
		for j := 0; j < n; j++ {
			presum[i+1][j+1] = presum[i+1][j] + presum[i][j+1] - presum[i][j] + img[i][j]
		}
	}
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			di, ui, dj, uj := max(0, i-1), min(m, i+2), max(0, j-1), min(n, j+2)
			ans[i][j] = (presum[ui][uj] - presum[ui][dj] - presum[di][uj] + presum[di][dj]) / ((ui - di) * (uj - dj))
		}
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 周赛/leetcode每日一题/_03_24/661. 图片平滑器.java
package leetcode每日一题._03_24;

class Solution {
    public int[][] imageSmoother(int[][] img) {
        int m = img.length, n = img[0].length;
        int[][] presum = new int[m + 1][n + 1], ans = new int[m][n];
        for(int i = 0; i < m; i++)
            for(int j = 0; j < n; j++)
                presum[i + 1][j + 1] = presum[i][j + 1] + presum[i + 1][j] - presum[i][j] + img[i][j];
        for(int i = 0; i < m; i++)
            for(int j = 0; j < n; j++) {
                int cnts = 0, downI = Math.max(0, i - 1), upI = Math.min(m, i + 2), downJ = Math.max(0, j - 1), upJ = Math.min(n, j + 2);
                for(int di = downI; di < upI; di++)
                    for(int dj = downJ; dj < upJ; dj++)
                        cnts++;
                ans[i][j] = (presum[upI][upJ] - presum[upI][downJ] - presum[downI][upJ] + presum[downI][downJ]) / cnts;
            }
        return ans;
    }
}



